### PR TITLE
For completions of union, exclude arrays

### LINKED ISSUE #13036: TypeScript intellisense does not work correctly for interface properties that can be an interface or an array of interfaces
_From @chase-smith on December 15, 2016 5:19_

- VSCode Version: 1.8.0
- OS Version: Linux Mint 17 Xfce 64-bit

TypeScript intellisense does not work correctly on interface properties that can be an interface or an array of that interface. Intellisense is only present when the property is in the array form. 

Code to reproduce:
```typescript
interface SomeInterface {
    prop1: number;
    prop2: string;
}
interface ComplexInterface {
    fail: SomeInterface | SomeInterface[];
}
let b:ComplexInterface = {
    fail: {}
}
```
Screenshot of what intellisense gives if the property is an object:
fail_pic_1

Screenshot of what intellisense gives if the property is an array of objects:
fail_pic_2

Expected behavior is that intellisense in the first screenshot would act like the second screenshot, and list properties `prop1` and `prop2`.

Edit: I found that if I use `fail: <SomeInterface>{}`, intellisense works as expected, as shown in the below screenshot. However, this seems like it shouldn't be necessary, given how good TypeScript is at inferring other things.
fail_pic_3

_Copied from original issue: Microsoft/vscode#17229_

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #13036

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
I don't really know about this change. Removing method-declared properties is a bit heavy-handed and likely to impact the experience negatively for a large class of users where `.d.ts` files have taken advantage of method syntax.

**Comment by andy-ms** [comment]
@DanielRosenwasser This only affects union types. In the usual case where you have just a single type, we will still show methods.

File: src/services/completions.ts [hunk 1]
```diff
@@ -1758,4 +1758,30 @@ namespace ts.Completions {
...
+                }
+                return memberType;
+            });
+            // If there are no property-only types, just provide completions for every type as usual.
+            if (propertyOnlyTypes.length > 0) {
+                type = checker.getUnionType(propertyOnlyTypes);
+            }
+        }
+
+        return checker.getAllPossiblePropertiesOfType(type);
```
Comment (mhegazy) [inline src/services/completions.ts]: can we instead have `getAllPossiblePropertiesOfTypes` that takes an array of types? i think this is the only reference to the internal API `getAllPossiblePropertiesOfType` anyways.

**Comment by andy-ms** [comment]
@DanielRosenwasser Any more comments?

**Comment by mhegazy** [comment]
What about excluding arrays only? i am having second thoughts here. Are Map/Set common? have you seen `PropertyBag | SomeOtherTye` pattern a lot in DT?

**Comment by DanielRosenwasser** [comment]
@mhegazy and I talked about just de-prioritizing items in the completion list from `Array`.

**Comment by mhegazy** [comment]
>  and I talked about just de-prioritizing methods in the completion list as well.

i do not think this is gonna work. for arrays, you do not want to see them period. you do not want to see `length` either..

Similarly for Map, and Set, having their methods in the list is just confusing regardless of search order.

But in the absence of a general solution that gets us `Map` and `Set`, we might just settle for `Array`.

**Comment by andy-ms** [comment]
@mhegazy @DanielRosenwasser OK, new version only deals with arrays (including types such as `ReadonlyArray` and `NodeArray`).

**Comment by mhegazy** [comment]
thanks!

**Pull request merged**